lib/checkother.cpp

```diff
@@ -1391,6 +1391,16 @@ void CheckOther::checkConstVariable()
         // Skip if address is taken
         if (Token::findmatch(var->nameToken(), "& %varid%", scope->bodyEnd, var->declarationId()))
             continue;
+        // Skip if another non-const variable is initialized with this variable
+        {
+            auto match = Token::findmatch(var->nameToken(), "%type% & %name% = %varid%", scope->bodyEnd, var->declarationId());
```

Comment:
An assignment to a reference should be checked in the `isVariableChanged` function instead.

Reply:
actually just to be clear, if x is the variable in question the expression we're detecting here is y = x, not x = y. Do you still think it should be considered modifying the variable? It does allow the variable to be modified but that's basically what all these checks are looking for.

Reply:
Actually, `isVariableChanged` already checks if its assigned to a non-const reference [here](https://github.com/danmar/cppcheck/blob/main/lib/astutils.cpp#L1477). The extra check here should not be necessary unless there is some bug in the code.

Reply:
That's not the case I'm testing for here (I think). Suppose I start with ```X& x```.
If I say ```x = _``` then isVariableChanged should (and does as far as I can tell) detect that x has been assigned to and thus cannot be const.

The case I'm trying to detect is shown here https://godbolt.org/z/ze8r8o
```xc``` cannot be const because it prevents y2 from compiling. More generally a reference from which a non-const reference is created cannot itself be const. Interestingly, I think I'm failing to respect the fact that a const& is still fine here. I'll add a test for that and attempt a fix.

Reply:
In that case `y2` can be `const`(and we should warn about this), and when that is changed to `const` then `xc` can be `const` as well. I dont really see that as an FP since we should warn about all the variables that can be declared `const`.

Reply:
that depends on how y2 is used. If we call a non-const method on y2 (as I do in the unit tests) then it can't be const. If we don't then it can and once we fix the ensuing warning we'll then get the warning about xc being const.

Reply:
>I think I'm failing to respect the fact that a const& is still fine here. I'll add a test for that and attempt a fix.

I was wrong about this. Looks like the test and code handle it.

Reply:
> If we call a non-const method on y2 (as I do in the unit tests) then it can't be const.

But this should be caught by `isVariableChanged`, if not there is a bug in `isVariableChanged`.

Reply:
should isVariableChanged return true if we call y2.mutate() and can't see the definition of y2.mutate()? i.e. does this method indicate we know the variable has changed or just that we can't know that it hasn't? I assumed the former.